symengine/basic.h

```diff
@@ -97,7 +97,15 @@ class Basic : public EnableRCPFromThis<Basic>
     mutable hash_t hash_; // This holds the hash value
 #endif // WITH_SYMENGINE_THREAD_SAFE
 public:
+#ifdef USE_VIRTUAL_TYPEID
     virtual TypeID get_type_code() const = 0;
+#else
+    TypeID type_code_;
+    inline TypeID get_type_code() const
+    {
+        return type_code_;
```

Comment:
I would also (in Debug mode) initialize `type_code_` to some bogus value (I think you can just use the following syntax `TypeID type_code_ = XXX;`, and then check in `get_type_code` that the `type_code_` is not equal to this bogus value.

Also, it would be worth exploring, whether you can't simply do:
```
const TypeID type_code_ = ID;
```
below. Wouldn't that work? That way you don't have to use that `ASSIGN_TYPEID()` at all.

If this doesn't work, then my other comment is that it's better (faster) I think to initialize such variables using the initializer list:  http://en.cppreference.com/w/cpp/language/initializer_list